### PR TITLE
1996 Grad Program

### PR DESCRIPTION
**Pull request by sreepadkarekkad**

### CODE REVIEW

File: api/src/main/java/ca/bc/gov/educ/api/graduation/model/report/ReportData.java [hunk 1]
```diff
@@ -1,96 +1,81 @@
...
+		@JsonSubTypes.Type(value = Certificate.class),
+		@JsonSubTypes.Type(value = OptionalProgram.class),
+		@JsonSubTypes.Type(value = GradRequirement.class),
+		@JsonSubTypes.Type(value = Exam.class),
+		@JsonSubTypes.Type(value = Assessment.class),
+		@JsonSubTypes.Type(value = AssessmentResult.class),
+		@JsonSubTypes.Type(value = AchievementCourse.class)
+})
+public class ReportData implements Serializable {
+
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/graduation/model/report/ReportData.java:50]: Hi Sreepad, can you implement serializable on all properties? For example, School, Transcript, GradProgram, etc should also implement serializable and have a 'private static final long serialVersionUID = {long}L' property for serialization compliance.

File: api/src/main/java/ca/bc/gov/educ/api/graduation/model/report/ReportRequest.java [hunk 2]
```diff
@@ -2,20 +2,30 @@
...
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+@JsonPropertyOrder({ "options", "data"})
+public class ReportRequest implements Serializable {
 
-    @JsonProperty("data")
-    public ReportData getData() { return reportData; }
-    public void setData(ReportData value) { this.reportData = value; }
-}
+	private ReportData data;
+	private ReportOptions options;
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/graduation/model/report/ReportRequest.java:15]: Hi Sreepad, also make ReportData and ReportOptions implement Serializable for serialization safety. Thanks.

File: api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java [hunk 3]
```diff
@@ -341,22 +330,23 @@ private GradStatus getGraduationStatus(ca.bc.gov.educ.api.graduation.model.dto.G
...
 		List<StudentCourse> studentCourseList = graduationDataStatus.getStudentCourses().getStudentCourseList();
 		List<StudentCourse> studentExamList = studentCourseList
 				.stream()
 				.filter(sc -> "Y".compareTo(sc.getProvExamCourse()) == 0)
 				.collect(Collectors.toList());
 		List<StudentAssessment> studentAssessmentList = graduationDataStatus.getStudentAssessments().getStudentAssessmentList();
-		List<StudCourse> sCourseList = new ArrayList<>();
-		List<StudAssessment> sAssessmentList = new ArrayList<>();
-		List<StudExam> sExamList = new ArrayList<>();
+		List<AchievementCourse> sCourseList = new ArrayList<>();
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java:340]: Remove sAssessmentList if not used.

File: api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java [hunk 4]
```diff
@@ -96,10 +83,10 @@ public ReportData prepareReportData(
 		return data;
 	}
 
-	private List<NonGradReason> getNonGradReasons(List<GradRequirement> nonGradReasons) {
+	private List<NonGradReason> getNonGradReasons(List<ca.bc.gov.educ.api.graduation.model.dto.GradRequirement> nonGradReasons) {
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java:86]: Code duplication: getNonGradReasons and getNonGradReasonsAchvReport (line 292) are identical. Can you remove one?

File: api/src/main/java/ca/bc/gov/educ/api/graduation/util/EducGraduationApiUtils.java [hunk 5]
```diff
@@ -107,6 +107,19 @@ public static String formatDateForReportJasper(String updatedTimestamp) {
 
 	}
 
+	public static Date formatIssueDateForReportJasper(String updatedTimestamp) {
+		SimpleDateFormat fromUser = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/graduation/util/EducGraduationApiUtils.java:111]: Consider using a constant for "yyyy-MM-dd" or a method for getting a SimpleDateFormat with a consistent type (like getSimpleDateFormat() which returns one formatted with "yyyy-MM-dd".

**Review by cditcher — CHANGES_REQUESTED** [review]
Hi Sreepad, just a couple of minor details. Please see comments. Thanks.

File: api/src/main/java/ca/bc/gov/educ/api/graduation/model/report/ReportData.java (hunk 1 above)
Comment (sreepadkarekkad) [reply api/src/main/java/ca/bc/gov/educ/api/graduation/model/report/ReportData.java:50]: I have made the changes

File: api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java (hunk 3 above)
Comment (sreepadkarekkad) [reply api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java:326]: Removed

File: api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java (hunk 4 above)
Comment (sreepadkarekkad) [reply api/src/main/java/ca/bc/gov/educ/api/graduation/service/ReportService.java:85]: Removed duplicate

File: api/src/main/java/ca/bc/gov/educ/api/graduation/util/EducGraduationApiUtils.java (hunk 5 above)
Comment (sreepadkarekkad) [reply api/src/main/java/ca/bc/gov/educ/api/graduation/util/EducGraduationApiUtils.java:111]: Moved to Utils

**Pull request merged**